### ISSUE TITLE
FileEventManager Events Should be Multi-File Events

### ISSUE THREAD

**Issue by ilexp** [AdamsLair/duality#323]
#### Summary

Currently, the global file system watcher events in the `FileEventManager` class deal with one path at a time. This severely limits scalability performance-wise. Instead, each of those events should deal with N files, and the `FileEventManager` should internally aggregate N consecutive events of the same kind to a single event.
#### How to reproduce
- Import 500 icons into the project and watch as the Project View updates for each added one individually. 
- Delete the 500 Resources and watch as it updates for each removed one individually.
#### Analysis
- There are cases where the order of events matters, so it will be important to not blindly aggregate all events of the same type, but to batch only _sequences_ of events of the same type.

**Comment by ilexp** [comment]
## Progress

- Started working on this in the `fileeventmanager-batching` branch.
- Refactored `FileEventManager` internals towards batching.
- Fixed a bug in global Resource directory rename operations when the renamed path contained repetitions, similar to the one fixed in `ContentProvider` in issue #343.

## ToDo

- Continue refactoring `FileEventManager` towards an internal "process a list of `FileEvent` structs" approach, avoiding allocation and usage of lots of `EventArgs` objects.
- Write unit tests for the basic functionality of `FileEventManager`, both system internals and global events. This will likely require some refactoring as well, since `FileEventManager` is currently hardwired to editor idle / form activate events. Consider reversing access patterns and providing a public API instead.
- Unify public Resource events into a single event providing a full list of change events in a single invocation.
- Manually test editor use cases for file events:
  - Auto re-import of assets on source file changes.
  - External changes in unsaved files.
  - External create, delete, rename of resource files in the Data folder.
  - External create, delete, rename of directories in the Data folder.
- Merge back to `develop-3.0`. Consider doing a squash-merge.

**Comment by ilexp** [comment]
## Progress

- More refactoring towards internal `FileEvent` struct usage in `FileEventManager`.
- Fixed an editor exception when attempting to auto-move a Source/Media directory containing locked files.
- Added path-only constructor for `ContentRef<T>`.

## ToDo

- Change file event processing into a three step process: Aggregate, Discard, Process. Each step should operate on the full list, which is transformed.
- Write unit tests for the basic functionality of `FileEventManager`, both system internals and global events. This will likely require some refactoring as well, since `FileEventManager` is currently hardwired to editor idle / form activate events. Consider reversing access patterns and providing a public API instead.
- Unify public Resource events into a single event providing a full list of change events in a single invocation.
- Manually test editor use cases for file events:
  - Auto re-import of assets on source file changes.
  - External changes in unsaved files.
  - External create, delete, rename of resource files in the Data folder.
  - External create, delete, rename of directories in the Data folder.
- Merge back to `develop-3.0`. Consider doing a squash-merge.

**Comment by ilexp** [comment]
## Progress

- Extracted `FileEvent`, `FileEventType` and `FileEventQueue` from the manager. They're also now core types, as they don't have any dependencies outside `Duality.IO`.
- Tweaked how events are aggregated and filtered.

## ToDo

- Write unit tests for `FileEventQueue` to check aggregation and filtering behavior.
- Improve aggregation and filtering behavior and actually track changes through the event queue to allow aggregating event chains of any depth.
- Write unit tests for the basic functionality of `FileEventManager`, both system internals and global events. This will likely require some refactoring as well, since `FileEventManager` is currently hardwired to editor idle / form activate events. Consider reversing access patterns and providing a public API instead.
- Unify public Resource events into a single event providing a full list of change events in a single invocation.
- Manually test editor use cases for file events:
  - Auto re-import of assets on source file changes.
  - External changes in unsaved files.
  - External create, delete, rename of resource files in the Data folder.
  - External create, delete, rename of directories in the Data folder.
- Merge back to `develop-3.0`. Consider doing a squash-merge.

**Comment by ilexp** [comment]
## Progress

- Added XML docs and extended usage of `FileEventQueue`.
- Added unit tests for basic `FileEventQueue` behavior, including basic aggregations and discards.

## ToDo

- Write unit tests for more complex `FileEventQueue` aggregation and discard scenarios.
  - Test the "Photoshop saved image" sequence of events, should be cleaned up to be in usable state.
  - Test "create, something, something, delete", should be discarded completely.
- Improve aggregation and filtering behavior and actually track changes through the event queue to allow aggregating event chains of any depth.
- Write unit tests for the basic functionality of `FileEventManager`, both system internals and global events. This will likely require some refactoring as well, since `FileEventManager` is currently hardwired to editor idle / form activate events. Consider reversing access patterns and providing a public API instead.
- Unify public Resource events into a single event providing a full list of change events in a single invocation.
- Manually test editor use cases for file events:
  - Auto re-import of assets on source file changes.
  - External changes in unsaved files.
  - External create, delete, rename of resource files in the Data folder.
  - External create, delete, rename of directories in the Data folder.
- Merge back to `develop-3.0`. Consider doing a squash-merge.

**Comment by ilexp** [comment]
## Progress

- Added advanced unit tests for `FileEventQueue`.
- Improved the aggregation algorithm to handle complex cases such as photoshop-like save overwrite with a lower amount of events.

## ToDo

- Write unit tests for the basic functionality of `FileEventManager`, both system internals and global events. This will likely require some refactoring as well, since `FileEventManager` is currently hardwired to editor idle / form activate events. Consider reversing access patterns and providing a public API instead.
- Unify public Resource events into a single event providing a full list of change events in a single invocation.
- Manually test editor use cases for file events:
  - Auto re-import of assets on source file changes.
  - External changes in unsaved files.
  - External create, delete, rename of resource files in the Data folder.
  - External create, delete, rename of directories in the Data folder.
- Merge back to `develop-3.0`. Consider doing a squash-merge.

**Comment by ilexp** [comment]
## Progress

- Introduced specialized EventArgs for quickly matching a list of file events against expected conditions.
- Unified public Resource events into a single event with a full list of changes.

## ToDo

- Introduce API for source directory events.
- Reconsider PluginChanged API to match or specialized.
- If possible within reasonable effort, write unit tests for the basic functionality of `FileEventManager`, both system internals and global events.
- Manually test editor use cases for file events:
  - Auto re-import of assets on source file changes.
  - External changes in unsaved files.
  - External create, delete, rename of resource files in the Data folder.
  - External create, delete, rename of directories in the Data folder.
  - Delete a Pixmap, observe CamView and Inspector changes to Material previews.
  - Rename a different Pixmap to match the deleted Pixmaps path, observe CamView and Inspector changes to Material previews.
- Merge back to `develop-3.0`. Consider doing a squash-merge.

**Comment by ilexp** [comment]
## Progress

- Introduced API for source change events.
- Changed API for plugin change events to match the new multi-event handling.
- Extracted asset re-import trigger on MainForm activate from FileEventManager.
- Exposed public API for immediately processing file events and asset re-imports.

## ToDo

- If possible within reasonable effort, write unit tests for the basic functionality of `FileEventManager`, both system internals and global events.
- Manually test editor use cases for file events:
  - Auto re-import of assets on source file changes.
  - Auto re-load of plugins on recompiles.
  - Auto load of plugins on first compile.
  - External changes in unsaved files.
  - Internal + External create, delete, rename of resource files in the Data folder.
  - Internal + External create, delete, rename of directories in the Data folder.
  - Delete a Pixmap, observe CamView and Inspector changes to Material previews.
  - Rename a different Pixmap to match the deleted Pixmaps path, observe CamView and Inspector changes to Material previews.
- Merge back to `develop-3.0`. Consider doing a squash-merge.

**Comment by ilexp** [comment]
## Progress

- Fixed a folder / file rename bug in Project View when it was still locked externally.
- Manually tested the following cases:
  - Auto re-import of assets on source file changes.
  - Auto re-load of plugins on recompiles.
  - Auto load of plugins on first compile.
  - External changes in unsaved files.
  - Internal + External create, delete, rename of resource files in the Data folder.
  - Internal + External create, delete, rename of directories in the Data folder.
  - Delete a Pixmap, observe CamView and Inspector changes to Material previews.

## ToDo

- Re-test the following case and ensure functionality: "Rename a different Pixmap to match the deleted Pixmaps path, observe CamView and Inspector changes to Material previews."
  - It worked so far, but renaming, deleting and creating don't seem to trigger editor wide change events themselves - only when they need to be propagated to dependent objects.
  - See if this can be improved upon.
- Check out how the EditorBasePlugin code for propagating changes to dependent objects can be optimized with the new multi-file API.
- Merge back to `develop-3.0`. Consider doing a squash-merge.

**Comment by ilexp** [comment]
## Progress

- FileEventManager now automatically transforms external Resource changes into ObjectPropertyChanged notifications, so plugins and editor modules no longer need to watch out for external and internal changes separately.
- Simplified plugin code to take advantage of the new external change transformation.
- Implicitly optimized external change propagation code (by replacing it with the above transformation), so it no longer handles every single external change separately.
- Re-tested the "rename X to deleted Y, observe changes to Y-dependent Z" case, now reacts immediately and schedules re-draws wherever necessary.

## ToDo

- Do some last manual tests, evaluate code, final tweaks where necessary.
- Merge back to `develop-3.0`. Consider doing a squash-merge.

**Comment by ilexp** [comment]
Squash-merged to `develop-3.0`. Full change log, since [commit](https://github.com/AdamsLair/duality/commit/39e0730d003c037f000c047d33d5714c91f94fa9) message rewrite was truncated for some reason:

```
FileEventManager Multi-File Events

#ADD: Added a new ContentRef shortcut constructor for refs based on paths only.
#ADD: Extracted FileEvent, FileEventType and FileEventQueue core types from the editor FileEventManager.
#ADD: Added automatest tests for FileEventQueue aggregation and discard behavior.
#ADD: Added EventArgs classes with specialized API to easily evaluate batches of FileEvents.
#ADD: Introduced FileEventManager SourcesChanged API
#ADD: External Resource changes are now translated into object change notification events by the FileEventManager.
#CHANGE: Simplified plugin code to react to Resource change notifications, regardless of their external or internal origin. (+23 squashed commit)
#CHANGE: Extracted Asset Re-Import Trigger from FileEventManager
#CHANGE: FileEventManager PluginsChanged Multi-Event API
#CHANGE: PluginsChanged is now a multi-event API similar to ResourcesChanged.
#CHANGE: Improved file event aggregation algorithm, so it can further simplify the event sequences generated by photoshop-like save / overwrite of files.
#CHANGE: Tabs for indentation, spaces for alignment.
#FIX: Fixed an editor exception when renaming a Data directory referring to locked Source/Media files.
#FIX: Fixed a bug in global directory rename when the renamed directory structure contained repetitions.
```
